sdks/go/pkg/beam/core/runtime/exec/sdf.go

```diff
@@ -457,6 +501,23 @@ type SplittableUnit interface {
 // each case occurs and the implementation details, see the documentation for
 // the singleWindowSplit and multiWindowSplit methods.
 func (n *ProcessSizedElementsAndRestrictions) Split(f float64) ([]*FullValue, []*FullValue, error) {
+	// Get the watermark state immediately so that we don't overestimate our current watermark.
+	var pWeState interface{}
+	var rWeState interface{}
```

Comment:
I'm not sure I fully understand your concern here - how are you suggesting we reduce decisions made? The only decision being made here is whether we need to invoke getWatermarkEstimatorState, which is a decision that should be made as close to when we actually perform the split as possible (without happening after the split to avoid accidental overestimation of state). Even if we know the restriction has watermark state, we will need this condition

Reply:
This was simplified by the default invoker approach.